operator/builtin/transformer/recombine/recombine.go

```diff
@@ -229,46 +259,59 @@ func (r *RecombineOperator) matchIndicatesLast() bool {
 }
 
 // addToBatch adds the current entry to the current batch of entries that will be combined
-func (r *RecombineOperator) addToBatch(_ context.Context, e *entry.Entry) {
-	if len(r.batch) >= r.maxBatchSize {
-		r.Error("Batch size exceeds max batch size. Flushing logs that have not been recombined")
-		r.flushUncombined(context.Background())
+func (r *RecombineOperator) addToBatch(_ context.Context, e *entry.Entry, source string) {
+	if _, ok := r.batchMap[source]; !ok {
+		r.batchMap[source] = []*entry.Entry{e}
+		if len(r.batchMap) >= r.maxSources {
+			r.Error("Batched source exceeds max source size. Flushing all batched logs. Consider increasing max_sources parameter")
+			r.flushUncombined(context.Background())
+		}
+		return
 	}
 
-	r.batch = append(r.batch, e)
+	r.batchMap[source] = append(r.batchMap[source], e)
+	if len(r.batchMap[source]) >= r.maxBatchSize {
+		if err := r.flushSource(source); err != nil {
+			r.Errorf("there was error flushing combined logs %s", err)
+		}
+	}
 }
 
 // flushUncombined flushes all the logs in the batch individually to the
 // next output in the pipeline. This is only used when there is an error
 // or at shutdown to avoid dropping the logs.
 func (r *RecombineOperator) flushUncombined(ctx context.Context) {
-	for _, entry := range r.batch {
-		r.Write(ctx, entry)
+	for source := range r.batchMap {
+		for _, entry := range r.batchMap[source] {
+			r.Write(ctx, entry)
+		}
```

Comment:
It would be nice if we combined entries for each source, but given that this is a fairly atypical bailout condition, I think it's ok to proceed with individual entries. I'll make a ticket to capture this as a future improvement.